src/cmdstan/command_helper.hpp

```diff
@@ -0,0 +1,589 @@
+  in.close();
+
+  size_t col_offset = 0;
+  for (auto name : names) {
+    if (boost::algorithm::ends_with(name, "__")) {
+      col_offset++;
+    } else {
+      break;
+    }
+  }
+  for (size_t i = 0; i < param_names.size(); ++i) {
+    if (param_names[i].compare(names[i + col_offset]) != 0) {
+      msg << "Mismatch between model params and StanCSV file \"" << fname
+          << "\",  expecting param \"" << param_names[i] << "\", found \""
+          << names[i + col_offset] << "\"" << std::endl;
+      throw std::invalid_argument(msg.str());
+    }
+    try {
+      theta_hat[i] = std::stod(values[i + col_offset]);
+    } catch (const std::exception &e) {
+      msg << "Error parsing CSV file, bad value for " << param_names[i]
+          << std::endl;
+      throw std::invalid_argument(msg.str());
+    }
+  }
+}
+
+/**
+ * Parse a JSON file and extract the model parameter estimates.
+ * Helper function get_var_context throws exception if cannot
+ * open or parse JSON file.
+ *
+ * @param fname name of file which exists and has read perms.
+ * @param param_names vector of model constrained parameter names.
+ * @param theta_hat Eigen vector for estimated parameters
+ */
+void get_theta_hat_json(const std::string &fname,
+                        const std::vector<std::string> &param_names,
+                        Eigen::VectorXd &theta_hat) {
+  std::vector<std::string> var_names;
+  std::vector<std::string> splits;
+  std::string cur_name = "";
+  for (auto &&param_name : param_names) {
+    boost::algorithm::split(splits, param_name, boost::is_any_of("."),
+                            boost::token_compress_on);
+    if (cur_name.compare(splits[0]) != 0)
+      var_names.push_back(splits[0]);
+    cur_name = splits[0];
+  }
+
+  std::shared_ptr<stan::io::var_context> context = get_var_context(fname);
+  Eigen::Index offset = 0;
+  for (auto &&var_name : var_names) {
+    const auto param_vec = context->vals_r(var_name);
+    for (Eigen::Index i = 0; i < param_vec.size(); ++i) {
+      theta_hat[offset] = param_vec[i];
+      ++offset;
+    }
+  }
+}
+
+/**
+ * Parse contents of file containing estimate of parameter modes.
+ *
+ * @param fname name of file which exists and has read perms.
+ * @param model Stan model
+ * @return Eigen vector of parameter estimates.
+ */
+Eigen::VectorXd get_laplace_mode(const std::string &fname,
+                                 const stan::model::model_base &model) {
+  std::stringstream msg;
+  std::vector<std::string> param_names;
+  get_constrained_params(model, param_names);
+  Eigen::VectorXd theta_hat(param_names.size());
+  if (suffix(fname) == ".csv") {
+    get_theta_hat_csv(fname, param_names, theta_hat);
+  } else if (suffix(fname) == ".json") {
+    get_theta_hat_json(fname, param_names, theta_hat);
+  } else {
+    msg << "Mode file must be CSV or JSON, found " << fname << std::endl;
+    throw std::invalid_argument(msg.str());
+  }
+  return theta_hat;
```

Comment:
Theta hat needs to be on the unconstrained scale (see non-review comment above or try a simplex model)

Reply:
fixed.